[HmuuMyatMoe] iP

HmuuMyatMoe opened the pull request:
# **Duke**

> “You were meant for greater things." 📈

Don't waste your time remembering tasks, let Duke do it for you. 👍 It is
- Easy to use
- Fast
- Convenient
- ****FREE****

## **Features** 📑
+ Add tasks
  + Categorize your tasks into Todo, Deadline or Event.
+ List out tasks
+ Mark tasks as done
+ Unmark tasks as done
+ Delete tasks
+ Find tasks

## **Usage guide** 📚
1. Download *"duke.jar"* file from [here](https://github.com/HmuuMyatMoe/ip/releases/tag/A-Jar).
2. Double-click the downloaded file.
3. Start adding & managing your tasks!

## **Try this** ⬇️
```
todo buy book
list
```

## Potential Upcoming Features 😉
[ ] Managing tasks
[ ] Managing deadlines
[ ] Reminders

## Program's main method [^1]
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

✨ Get DUKE today! ✨

[^1]: TIP! If you are a java programmer, feel free to practice your Java programming using this.

Commit:
Add Gradle support

HmuuMyatMoe committed:
Level-1

HmuuMyatMoe committed:
added list & add function & print function

HmuuMyatMoe committed:
created Task class, added 'mark as done' function & 'unmark' function

HmuuMyatMoe committed:
added todo, deadline & event

HmuuMyatMoe committed:
edited print statements & added test cases

HmuuMyatMoe committed:
handle invalid input and incomplete desc errors

HmuuMyatMoe committed:
delete function, OutOfBoundsException

HmuuMyatMoe committed:
enum

HmuuMyatMoe committed:
changed the add task function & updated test cases

HmuuMyatMoe committed:
save the storage data in hard disk

HmuuMyatMoe committed:
Stores deadlines and event dates as java.time.LocalDate

Stored deadlines and event dates as Strings

Teach chatbot to understand dates instead of just storing them as Strings. Formatted dates are more organised and easier for users to understand.

HmuuMyatMoe committed:
Merge branch 'master' into branch-Level-8

HmuuMyatMoe committed:
Deadline & Event stores dates as LocalDate.
Date is stored and read in yyyy/MM/dd format and printed out to user in "day month year" format.

Dates are printed as such so that they are easier for users to understand.

HmuuMyatMoe committed:
Extracted out the classes - Ui, Parser, Command, Add Command, Delete Command, Mark Command, Unmark Command, List Command, Bye Command, Formatter

Encapsulated related codes in one class to follow good OOP practices.

HmuuMyatMoe committed:
Organise classes into appropriate packages.

Classes were not organised into packages.

Oragnaising the classes makes the code neater and more maintainable.

HmuuMyatMoe committed:
Automate project builds using gradle.

HmuuMyatMoe committed:
Add JUnit tests to test the behavior of the code for Task, ToDo, Deadline & Event classes.

Did not have JUnit tests and was not able to do unit testing.

JUnit tests allow for automated unit testing which is ensures each code works correctly.

HmuuMyatMoe committed:
Package the application as an executable JAR file

No JAR file to be distributed.

Package the app as a JAR file so that it can be distributed and users can run the JAR file.

HmuuMyatMoe committed:
Add JavaDoc comments for all classes and methods

Some classes and methods do not have JavaDoc comments

Adding JavaDoc comments make it easier for others to understand the code

HmuuMyatMoe committed:
Update the code to follow the specified Java coding standard given for this project

Some codes did not comply with the coding standard

Following the coding standard makes the code neater and easier for others to understand.

HmuuMyatMoe committed:
Fix checkstyle errors for JavaDoc comments

Some comments were not following the coding standard

Following coding standard makes code neater and easier to understand

HmuuMyatMoe committed:
Fix checkstyle errors for Tests

Some test codes had checkstyle error

Complying with coding standards make code more readable and easier to maintain

HmuuMyatMoe committed:
Resolve merge conflicts

Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Formatter.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/TaskList.java

HmuuMyatMoe committed:
Resolve merge conflicts in TaskList

Merge master with branch-A-CodingStandard which resulted in merge conflits

HmuuMyatMoe committed:
Add find function which allows user to search for a task using a keyword

There were no find function before

Adding the find function increases the quality of the program

Let's
* add a find function
* make the find command case insensitive

HmuuMyatMoe committed:
Resolve merch conflicts

Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/task/TaskList.java

RubyNguyen07 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,203 @@
+     * @throws InvalidInputException If due date/time is given in the wrong format.
+     */
+    private static Deadline getDeadline(String[] splitInputs) throws IncompleteDescException, InvalidInputException {
+        if (splitInputs.length <= 1) {
+            throw new IncompleteDescException(
+                    "The description of a deadline cannot be empty!");
+        }
+        int endIndex = splitInputs[1].indexOf("/by");
+        if (endIndex < 0) {
+            throw new IncompleteDescException(
+                    "Please add the due date/time!");
+        }
+        String name = splitInputs[1].substring(0, endIndex).strip();
+        String end = splitInputs[1].substring((endIndex + 3)).strip();
+        if (name.isBlank()) {
+            throw new IncompleteDescException(
+                    "The description of a deadline cannot be empty!");
+        }
+        if (end.isBlank()) {
+            throw new IncompleteDescException(
+                    "Please add the due date/time!");
+        }
+        try {
+            LocalDate endLd = Parser.parseDate(end);
+            end = Formatter.formatDateForPrint(endLd);
+            if (endLd.isBefore(LocalDate.now())) {
+                throw new InvalidInputException(
+                        "The given deadline (yyyy/mm/dd) " + end + " has passed.");
+            }
+            return new Deadline(name, endLd);
+        } catch (DateTimeParseException e) {
+            throw new InvalidInputException(
+                    "Please enter the date/time in \"yyyy/mm/dd\" format.");
+        }
```
Do you think dividing logical units into separate blocks would improve readability? 🤔

RubyNguyen07 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,203 @@
+     */
+    private static Event getEvent(String[] splitInputs) throws IncompleteDescException, InvalidInputException {
+        if (splitInputs.length <= 1) {
+            throw new IncompleteDescException(
+                    "The description of an event cannot be empty!");
+        }
```
Line break does help make it easier to see!

RubyNguyen07 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,203 @@
+                    "Please add the due date/time!");
+        }
+        try {
+            LocalDate endLd = Parser.parseDate(end);
```
Do you think a comment explaining what endLd is will make the code clearer?

RubyNguyen07 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,118 @@
+            Scanner sc = new Scanner(myFile);
+            while (sc.hasNextLine()) {
+                String data = sc.nextLine();
+                String[] dataArr = data.split(" \\| ");
```
As the type here is String[], do you think a plural name will be more suitable?

RubyNguyen07 reviewed (commented):
LGTM! Just a few comments to take into consideration.

HmuuMyatMoe committed:
Add javadoc comments for Junit tests
Run checkstyle tests

No javadoc comments for Junit tests

Let's
- add java doc comments for Junit tests
- run checkstyle tests

HmuuMyatMoe committed:
Implement a Graphical User Interface (GUI)

Use text-based user interface which is less visually appealing than a GUI

A GUI is more visually appealing and could be more user-friendly

Let's use JavaFx to implement a GUI

HmuuMyatMoe committed:
Use Varargs in Launcher and Duke

Use String[] and ArrayList which does not accept a flexible number of arguments

Varargs reduces the need for overloading

Let's use varargs where applicable

Sheemo commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+     *
+     * @return True if the program should exit, false otherwise.
+     */
+    public abstract boolean isExit();
```
isExit() could return false by default then only be overriden in ByeCommand to return true so you don't need to implement it in all your classes :)

Sheemo commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,203 @@
+    protected enum CommandEnum {
+        BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, FIND
+    }
+
```
Nice use of enum for the commands

Sheemo commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,116 @@
+     * @param tasks Varargs of tasks.
+     */
+    public void write(Task... tasks) {
+        try {
```
Nice use of varargs!

HmuuMyatMoe committed:
Package the application updated with GUI as an executable JAR file

There is a JAR file build using code that implemented Text UI

Packaging the new code that implements GUI to release the latest version
of the application

Let's package the new version of the app as a JAR file
so that it can be distributed and users can run the JAR file.

HmuuMyatMoe committed:
Implement Single Level of Abstraction Principle (SLAP) for Parser

Varying levels of abstraction within a code fragment reduces code
readability and makes it harder to maintain code

SLAP is followed to ensure that the level of code abstractions in Parser
is the same

Let's abstract code snippets into functions

so that we can level the abstraction level within a code snippet

HmuuMyatMoe committed:
Add Java assertions to define assumptions about the program state

Assumptions not checked

Not checking assumptions could result in bugs and cause program to run
in an undesired way

Let's
- Use assertions to check the program state at different points of
the program
- Check internal invariants and post conditions

Using assertions ensures that assumptions in the code are verified
during run time and that program terminates if a bug is detected
This helps to detect bugs and improve code

HmuuMyatMoe committed:
Refactor code

Some code snippets had varying levels of abstractions.
Some codes unused, some duplicated.

Unifying repeated code, deleting unused code
and maintaining single-level of abstraction in a code segment improves code quality, and makes code easier to maintain.

Let's
* Unify repeated code by abstracting out as functions
* Find and remove unused code
* Maintain single level of abstraction in a code segment

HmuuMyatMoe committed:
Use streams to write data to storage file in hard disk

For loop use to write string

Streams makes code more sophisticated and easier to read

Let's use streams where applicable

HmuuMyatMoe committed:
Merge pull request #1 from HmuuMyatMoe/branch-A-Assertions

Add Java assertions to define assumptions about the program state

HmuuMyatMoe committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java

HmuuMyatMoe committed:
Merge branch 'master' into branch-A-CodeQuality

HmuuMyatMoe committed:
Merge pull request #2 from HmuuMyatMoe/branch-A-CodeQuality

Refactor code

HmuuMyatMoe committed:
Merge branch 'master' into branch-A-CodeQuality

HmuuMyatMoe committed:
Revert "Refactor code"

HmuuMyatMoe committed:
Merge pull request #3 from HmuuMyatMoe/revert-2-branch-A-CodeQuality

Revert "Refactor code"

HmuuMyatMoe committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	data/DukeList.txt

HmuuMyatMoe committed:
Reset data file

HmuuMyatMoe committed:
Merge pull request #4 from HmuuMyatMoe/branch-A-CodeQuality

Refactor code

HmuuMyatMoe committed:
Merge branch 'master' into branch-A-Streams

HmuuMyatMoe committed:
Merge pull request #5 from HmuuMyatMoe/branch-A-Streams

Merge branch-A-Streams

HmuuMyatMoe committed:
Set up GitHub Continuous Integration (CI)

No github continuous integration.

Setting up CI allows for automation of workflow.

Let's set up CI and create workflow for push and pull requests.

HmuuMyatMoe committed:
Add two more types of find (find tasks with names that partially matches
the given keyword and find tasks using dates)

There is only one find command that finds tasks with names that
completely matches the given keyword which limits the
functionality of the program.

Adding more ways users can use to find tasks using different
keywords makes program more user-friendly and useful.

To give users more flexibility in searching tasks, let's
* add a findflex command that allows users to find tasks with names that
  partially matches the given keyword
* add a finddate command that helps users find tasks using a date given

HmuuMyatMoe committed:
Add more JUnit tests for Deadline, Event and ToDo
Made Task abstract

Task not being abstract allows for instantiation of a Task object which
we do not want to allow.

Task made abstract to prevent instantiation
More JUnit tests added to test codes automatically.

Let's
* add more JUnit tests
* make Task abstract

HmuuMyatMoe committed:
Change name and modify phrases used by chat bot

Duke chatbot personality is not unique.

Giving chatbot a unique personality differentiates it from other
chatbots and gives users a better experience.

Let's give a unique personality by
* changing the name
* modifying the phrases used by chatbot

HmuuMyatMoe committed:
Improve GUI design for better user experience

Simple and plain text box design with no borders and small text
can make it hard to see and differentiate user's and bot's texts.
White background can make application look dull.

Added border and differentiating user's and bot's texts
using different background colors makes it easier for user to
differentiate the texts.
Adding colors also makes the program more fun to use.

To improve user experience, let's
* add borders to the dialog boxes
* add background color to the dialog boxes

HmuuMyatMoe committed:
Abstract out a Gui class which is responsible for generating
responses to user

Without the Gui class, chat bot's responses to users were generated in
different methods, which can lead to inconsistencies.

Abstracting the gui class out prevents inconsistencies and minimise
repetitions.

Let's abstract out the gui class and methods that generate responses to
users.

HmuuMyatMoe committed:
Merge branch 'branch-A-MoreTesting'

# Conflicts:
#	src/test/java/botanic/task/TaskTest.java

HmuuMyatMoe committed:
Fix check style errors

Check style errors means code does not follow the coding standards
and may be hard for others to understand.

Following coding standards and fixing check style errors make code
neater and easier to understand.

Let's fix check style errors in the code.

HmuuMyatMoe committed:
Modify and ensure code and comments follows SE-EDU Java coding standard
Give credit to resources used and referenced in project

Code not following coding standard could make it hard for others to read
and understand.

Following coding standard ensures that code is near and easier to read
and maintain.
Giving due credit acknowledges other's work.

Let's
* follow java coding standard
* give due credit and attribution for reuse

HmuuMyatMoe committed:
Add screenshot showing product Ui
Add user guide

Without user guide or product ui screenshot,
users will not know what to except
and will not know how to use the program.

Adding a product Ui screenshot and a user guide ensures that users
are able to learn how to use the application easily, and encourages
users to use the application.

Let's
* add an image that showcases the product's Ui
* add a comprehensive user guide

HmuuMyatMoe committed:
Change order of import statements in JUnit test files
Update README to look neater for user readability

Inconsistent order of import statements can make code look messy.
Chunks of sentences in user guide can be hard to read.

Ensuring that order of import statements is consistent make code neater
and follows Java coding standard.
Separating chunks of sentences using line spaces increases readability.

Let's
* ensure consistent ordering of import statements
* avoid chunks of sentences in user guide

HmuuMyatMoe committed:
Update credits given for reuse

Not giving due credit is disrespectful to the source owner.

Giving credit acknowledges the work of the original source owner.

Let's make sure to give credit for reuse and referencing.
